lib/benchmark/benchmark.go

```diff
@@ -0,0 +1,119 @@
+package benchmark
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/HdrHistogram/hdrhistogram-go"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/trace"
+	logrus "github.com/sirupsen/logrus"
+)
+
+const (
+	// PauseTimeBetweenBenchmarks is the time to pause between each benchmark
+	PauseTimeBetweenBenchmarks = time.Second * 5
+)
+
+// Generator provides a standardized way to get successive benchmarks from a consistent interface
+type Generator interface {
+	// Generator advances the Generator to the next generation.
+	// It returns false when the generator no longer has configurations to run.
+	Generator() bool
+	// GetBenchmark returns the benchmark config for the current generation.
+	// If called after Generate() returns false, this will result in an error.
+	GetBenchmark() (context.Context, Config, error)
+	// Benchmark runs the benchmark of receiver type
+	// return an array of Results that contain information about the generations
+	Benchmark([]string, *client.TeleportClient) ([]*Result, error)
+}
+
+// Run is used to run the benchmarks, it is given a generator, command to run,
+// a host, host login, and proxy. If host login or proxy is an empty string, it will
+// use the default login
+func Run(cnf interface{}, cmd, host, login, proxy string) ([]Result, error) {
+	var results []Result
+	tc, err := makeTeleportClient(host, login, proxy)
+	if err != nil {
+		return nil, err
+	}
+	logrus.SetLevel(logrus.ErrorLevel)
+	command := strings.Split(cmd, " ")
+
+	switch c := cnf.(type) {
+	case *Linear:
+		c.config.Command = command
+		results, err = c.Benchmark(context.Background(), command, tc)
+		if err != nil {
+			return results, err
+		}
+		return results, nil
+	}
```

Comment:
I run into the issue of then having to check what generator type inside benchmark as well. If I were to do this like you suggested, it just doesn't really simplify things.

```
func (g *Generator) Benchmark(ctx context.Context, command []string, tc *client.TeleportClient) ([]Result, error) {
      // check type and convert, we need to check because all types have different inner methods
       lin  := Linear{} // convert to type
	var result Result
	var results []Result
	sleep := false

	for lin.Generate() { 
		ctx, cancel := context.WithCancel(ctx)
		defer cancel()
    ...
}
```